File: sao/src/bus.js
/* This file is part of Tryton.  The COPYRIGHT file at the top level of
   this repository contains the full copyright notices and license terms. */
(function() {
    'use strict';

    Sao.Bus = {};

    // Bus Identifier
    Sao.Bus.id = Sao.common.uuid4();
    Sao.Bus.channel_actions = {};
    Sao.Bus.request = null;
    Sao.Bus.last_message = undefined;
    Sao.Bus.listening = false;

    Sao.Bus.listen = function(last_message, wait) {
        wait = wait || 1;
        last_message = last_message || Sao.Bus.last_message;
        var session = Sao.Session.current_session;
        if (!session || !session.bus_url_host) {
            return;
        }
        Sao.Bus.listening = true;

        let channels = Object.keys(Sao.Bus.channel_actions);
        let url = new URL(`${session.database}/bus`, session.bus_url_host);
        Sao.Bus.last_message = last_message;
        Sao.Bus.request = jQuery.ajax({
            headers: {
                Authorization: 'Session ' + session.get_auth(),
            },
            contentType: 'application/json',
            data: JSON.stringify({
                last_message: last_message,
                channels: channels,
            }),
            dataType: 'json',
            url: url,
            type: 'POST',
            timeout: Sao.config.bus_timeout,
        });

        Sao.Bus.request.done(function(response) {
            if (Sao.Session.current_session != session) {
                return;
            }
            if (response.message) {
                last_message = response.message.message_id;
                Sao.Logger.info(
                    "poll channels %s with last message",
                    Sao.Bus.channels, last_message);
                Sao.Bus.handle(response.channel, response.message);
            }
            Sao.Bus.listen(last_message, 1);
        });

        Sao.Bus.request.fail(function(response, status, error) {
            if (Sao.Session.current_session != session) {
                return;
            }
            if ((error == "abort") || (error === "timeout")) {
                Sao.Bus.listen(last_message, 1);
            } else if (response.status == 501) {
                Sao.Logger.info("Bus not supported");
                Sao.Bus.listening = false;
            } else {
                Sao.Logger.error(
                    "An exception occured while connection to the bus. " +
                    "Sleeping for %s seconds",
                    wait, error);
                Sao.Bus.listening = false;
                window.setTimeout(
                    Sao.Bus.listen,
                    Math.min(wait * 1000, Sao.config.bus_timeout),
                    last_message, wait * 2);
            }
        });
    };

    Sao.Bus.handle = function(channel, message) {
        let actions = Sao.Bus.channel_actions[channel] || [];
        for (let callback of actions) {
            callback(message);
        }
    };

    Sao.Bus.register = function(channel, func) {
        if (!Object.hasOwn(Sao.Bus.channel_actions, channel)) {
            Sao.Bus.channel_actions[channel] = [];
        }
        Sao.Bus.channel_actions[channel].push(func);

        if (Sao.Bus.request) {
            Sao.Bus.request.abort();
        }
    };

    Sao.Bus.unregister = function(channel, func) {
        if (Object.hasOwn(Sao.Bus.channel_actions, channel)) {
            let actions = Sao.Bus.channel_actions[channel];
            let func_idx = actions.indexOf(func);
            if (func_idx != -1) {
                actions.splice(func_idx, 1);
            }
        }

        if ((Sao.Bus.channel_actions[channel] || []).length == 0) {
            delete Sao.Bus.channel_actions[channel];
        }
    }

    let popup_notification = function(message) {
        if (message.type != 'notification') {
            return;
        }

        try {
            if (Notification.permission != "granted") {
                return;
            }
        } catch (e) {
            Sao.Logger.error(e.message, e.stack);
            return;
        }

        new Notification(message.title, {
            body: message.body || '',
        });
    }
    Sao.Bus.register(`client:${Sao.Bus.id}`, popup_notification);

}());

File: tryton/tryton/bus.py
# This file is part of Tryton.  The COPYRIGHT file at the top level of
# this repository contains the full copyright notices and license terms.
import base64
import json
import logging
import socket
import threading
import time
import uuid
from collections import defaultdict
from urllib.error import HTTPError
from urllib.request import Request, urlopen

from gi.repository import GLib

from tryton.config import CONFIG
from tryton.jsonrpc import object_hook

logger = logging.getLogger(__name__)


class Bus:

    ID = str(uuid.uuid4())
    current_thread = None
    channel_actions = defaultdict(list)
    listening = False

    @classmethod
    def listen(cls, connection):
        if not CONFIG['thread']:
            return
        listener = threading.Thread(
            target=cls._listen, args=(connection,), daemon=True)
        listener.start()
        cls.current_thread = listener.ident

    @classmethod
    def _listen(cls, connection):
        bus_timeout = CONFIG['client.bus_timeout']
        session = connection.session
        authorization = base64.b64encode(session.encode('utf-8'))
        headers = {
            'Content-Type': 'application/json',
            'Authorization': b'Session ' + authorization,
            }

        thread_id = threading.get_ident()
        wait = 1
        last_message = None
        url = None
        while connection.session == session:
            if url is None:
                if connection.url is None:
                    time.sleep(1)
                    continue
                url = connection.url + '/bus'
            cls.listening = True

            channels = list(cls.channel_actions)
            request = Request(url,
                data=json.dumps({
                        'last_message': last_message,
                        'channels': channels,
                        }).encode('utf-8'),
                headers=headers)
            logger.info('poll channels %s with last message %s',
                channels, last_message)
            try:
                response = urlopen(request, timeout=bus_timeout)
                wait = 1
            except socket.timeout:
                wait = 1
                continue
            except Exception as error:
                if thread_id != cls.current_thread:
                    break
                if isinstance(error, HTTPError):
                    if error.code in (301, 302, 303, 307, 308):
                        url = error.headers.get('Location')
                        continue
                    elif error.code == 501:
                        logger.info("Bus not supported")
                        break
                logger.error(
                    "An exception occurred while connecting to the bus. "
                    "Sleeping for %s seconds",
                    wait, exc_info=error)
                cls.listening = False
                time.sleep(min(wait, bus_timeout))
                wait *= 2
                continue

            if connection.session != session:
                break
            if thread_id != cls.current_thread:
                break

            data = json.loads(response.read(), object_hook=object_hook)
            if data['message']:
                last_message = data['message']['message_id']
                GLib.idle_add(cls.handle, data['channel'], data['message'])
        cls.listening = False

    @classmethod
    def handle(cls, channel, message):
        for callback in cls.channel_actions[channel]:
            callback(message)

    @classmethod
    def register(cls, channel, function):
        from tryton import rpc

        restart = channel not in cls.channel_actions
        cls.channel_actions[channel].append(function)
        if restart:
            # We can not really abort a thread, so we will just start a new one
            # and ignore the result of the one already running
            Bus.listen(rpc.CONNECTION)

    @classmethod
    def unregister(cls, channel, function):
        try:
            cls.channel_actions[channel].remove(function)
        except ValueError:
            pass

        if not cls.channel_actions[channel]:
            del cls.channel_actions[channel]


def popup_notification(message):
    if message['type'] != 'notification':
        return

    from tryton.gui.main import Main
    app = Main()
    app.show_notification(
        message.get('title', ''), message.get('body', ''),
        message.get('priority', 1))


Bus.register(f'client:{Bus.ID}', popup_notification)
